### PR TITLE
Implement new definitions

### PR DESCRIPTION
**Pull request by omarudolley**
Implement new definitions
- Employment/WorkContract
- Permits/WorkPermit
- Employment/IncomeTax

### CODE REVIEW

File: src/Employment/WorkContractRequest_v1.0.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        description="The street name of the employer contact address",
+        examples=["Ruoholahdenkatu 17 A 6"],
+    )
+    postal_code: str = Field(
+        ...,
+        title="Postal Code",
+        max_length=250,
+        description="The postal code of the employer address",
+        examples=["00180"],
+    )
```
Comment (lietu) [inline src/Employment/WorkContractRequest_v1.0.py:44]: Do we really want `Postal code` to be 250 characters and `Street name` to be 10?

Comment (lietu) [reply src/Employment/WorkContractRequest_v1.0.py:44]: I believe the example doesn't fit in "street name" validations

File: src/Permits/WorkPermit_v1.0.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    A = "A"
+    B = "B"
+
+
+class Permit(CamelCaseModel):
+    permit_name: str = Field(
+        ...,
+        title="Permit Name",
+        description="The name of the permit issued for a person for residence ",
+        examples=["Seasonal work certificate "],
```
Comment (lietu) [inline src/Permits/WorkPermit_v1.0.py:20]: is the trailing space intentional on the above two fields?

**Comment by joakimnordling** [comment]
According to the task all the new definitions were supposed to be version `0.1`. These seem to be `1.0`.

File: src/Employment/WorkContractRequest_v1.0.py
Comment (joakimnordling) [inline src/Employment/WorkContractRequest_v1.0.py:1]: This was supposed to be called `Employment/WorkContract`, not `Employment/WorkContractRequest`

File: src/Employment/WorkContractRequest_v1.0.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+from datetime import date
+from enum import Enum
+from typing import List, Optional
+
+from definition_tooling.converter import CamelCaseModel, DataProductDefinition
+from pydantic import Field
+
+
+class WorkContractRequest(CamelCaseModel):
+    end_user_authentication_token: str = Field(
```
Comment (joakimnordling) [inline src/Employment/WorkContractRequest_v1.0.py:11]: Why is this here still after this comment?
image
So the only input is the authentication token (that's gets enabled when you mark the definition with `requires_authorization=True,`, as is done, and there's not supposed to be any input field in the request.

File: DataProducts/Employment/WorkContractRequest_v1.0.json [hunk 4]
```diff
@@ -0,0 +1,877 @@
...
+      "EmployerInfo": {
+        "properties": {
+          "name": {
+            "type": "string",
+            "maxLength": 250,
+            "title": "Name",
+            "description": "The official name of the employer",
+            "examples": ["Staffpoint Oy"]
+          },
+          "businessId": {
```
Comment (joakimnordling) [inline DataProducts/Employment/WorkContractRequest_v1.0.json]: In the spreadsheet this is written as `businessID`.

File: src/Employment/WorkContract_v1.0.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        ...,
+        title="Business ID",
+        max_length=250,
+        description="The business ID of the employer",
+        examples=["2492090-1"],
+    )
+    street_name: str = Field(
+        ...,
+        title="Street Name",
+        max_length=250,
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py:34]: This does not match with the spreadsheet, it says 40.

File: src/Employment/WorkContract_v1.0.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        ...,
+        title="Postal Code",
+        max_length=10,
+        description="The postal code of the employer address",
+        examples=["00180"],
+    )
+    city: str = Field(
+        ...,
+        title="City",
+        max_length=250,
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py:48]: Max length does not match with the spreadsheet.

File: src/Employment/WorkContract_v1.0.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+        ...,
+        title="City",
+        max_length=40,
+        description="The city of the employee address",
+        examples=["London"],
+    )
+    country: Optional[str] = Field(
+        ...,
+        title="Country",
+        max_length=3,
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py:92]: If we want this to be a 3 letter code for the country, should it possibly have both min_length and max_length set to 3?

File: src/Employment/WorkContract_v1.0.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+    )
+    city: str = Field(
+        ...,
+        title="City",
+        max_length=40,
+        description="The city of the employee address",
+        examples=["London"],
+    )
+    country: Optional[str] = Field(
+        ...,
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py:90]: Should the default value be `None` if this is an optional field?

File: src/Employment/WorkContract_v1.0.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+        None,
+        title="Termination",
+        max_length=500,
+        description="The terms and conditions for the contract termination",
+        examples=[
+            "To terminate the work contract, the terminating party must provide 30 days of written notice to "
+            "the other party"
+        ],
+    )
+    other_terms: Optional[List[Term]] = Field(
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py:325]: We've usually not made lists Optional, rather kept them required, but they can of course be empty lists. I updated the spreadsheet with Henna, so this field is now required (can be empty list).

File: src/Employment/WorkContract_v1.0.py [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+        max_length=10,
+        description="The work industry based on the Statistical classification of economic activities in the European "
+        "Community, abbreviated as NACE.",
+        examples=["79.1"],
+    )
+    locations: List[str] = Field(
+        None,
+        title="Locations",
+        description="The region in Finland where the work is done",
+        examples=["Rovaniemi", "Lapland"],
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py:150]: This is supposed to be one example.
```suggestion
        examples=["Rovaniemi, Lapland"],
```

File: src/Employment/WorkContract_v1.0.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+
+
+class PaymentGrounds(str, Enum):
+    MONTHLY = "monthly"
+    HOURLY = "hourly"
+
+
+class Compensation(CamelCaseModel):
+    payment_grounds: PaymentGrounds = Field(
+        None,
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py]: For a required field this should be `...`.

File: src/Employment/WorkContract_v1.0.py [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+        title="Payment Time",
+        description="The conditions for paying the salary",
+        examples=["The 15th of every month"],
+        max_length=250,
+    )
+
+
+class BenefitType(str, Enum):
+    PART_OF_SALARY = "part of salary"
+    ADDITIONAL_TO_SALARY = "additional to salary"
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py]: image

File: src/Employment/WorkContract_v1.0.py [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+    )
+
+
+class DeterminationOfHoliday(str, Enum):
+    ANNUAL_HOLIDAY_ACT = "annual holiday act"
+    COLLECTIVE_AGREEMENT = "collective agreement"
+
+
+class Holidays(CamelCaseModel):
+    paid_holiday_included: bool = Field(
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py]: Wrong name on field:
image

File: src/Employment/WorkContract_v1.0.py [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+        ...,
+        title="Paid Holiday",
+        description="Indicates if the contract includes paid holiday",
+        examples=[True],
+    )
+    number_of_holidays: Optional[int] = Field(
+        None,
+        title="Number of Holidays",
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py]: image

File: src/Employment/WorkContract_v1.0.py [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+        description="The terms used for the paid holidays",
+        examples=[5],
+    )
+    determination_of_holiday: Optional[DeterminationOfHoliday] = Field(
+        None,
+        title="Determination of Holiday",
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v1.0.py]: image

File: src/Employment/WorkContract_v1.0.py (hunk 12 above)
Comment (joakimnordling) [reply src/Employment/WorkContract_v1.0.py]: Never mind, discussed with Henna, agreed to keep them with spaces.

File: src/Employment/WorkContract_v0.1.py [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+        title="Payment Time",
+        description="The conditions for paying the salary",
+        examples=["The 15th of every month"],
+        max_length=250,
+    )
+
+
+class BenefitType(str, Enum):
+    PART_OF_SALARY = "partOfSalary"
+    ADDITIONAL_TO_SALARY = "additionToSalary"
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v0.1.py:219]: Sorry, checked with Henna these and it was an error these were camelCased in the spreadsheet. They're now snake_cased in the spreadsheet, need to be reverted to that here as well.

File: src/Employment/WorkContract_v0.1.py [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+
+class EmployerInfo(CamelCaseModel):
+    name: str = Field(
+        ...,
+        title="Name",
+        max_length=250,
+        description="The official name of the employer",
+        examples=["Staffpoint Oy"],
+    )
+    businessID: str = Field(
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v0.1.py:22]: Can this capitalization not be achieved using an alias or something like that?

File: src/Employment/WorkContract_v0.1.py [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+        ...,
+        title="Name",
+        max_length=250,
+        description="The name of the employee",
+        examples=["Tom Williams"],
+    )
+    street_address: str = Field(
+        ...,
+        title="Street Address",
+        max_length=40,
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v0.1.py:69]: Wrong max_length.

File: src/Employment/WorkContract_v0.1.py [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+
+
+class TermsOfWork(CamelCaseModel):
+    employment_start: date = Field(
+        ...,
+        title="Employment Start",
+        description="The start date of the employment",
+        examples=[datetime.datetime(2023, 11, 7)],
+    )
+    employment_end: date = Field(
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v0.1.py:111]: I realized that if this is a mandatory field, then you can only have fixed term contracts which was not the idea. Checked with Henna and updated this to an optional field, so please change accordingly.

File: src/Employment/WorkContract_v0.1.py [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+    determination_of_holiday: Optional[DeterminationOfHoliday] = Field(
+        None,
+        title="Determination Of Holiday",
+        description="The determination of the holiday",
+        examples=["The holiday is determined by the employer"],
+    )
+
+
+class Term(CamelCaseModel):
+    term_description: Optional[str] = Field(
```
Comment (joakimnordling) [inline src/Employment/WorkContract_v0.1.py:272]: It makes no sense for this to be optional. You would then have a list of `Term`s, where some have a `term_description` (only thing in it) set to `null`. That's not desirable, rather the item should be left out in that case. Checked this with Henna and updated spreadsheet, so please update this to be a required field.

File: src/Permits/WorkPermit_v0.1.py [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+        description="The start date of the residence permit validity",
+        examples=[datetime.datetime(2023, 11, 7)],
+    )
+    validity_end: Optional[date] = Field(
+        None,
+        title="Validity End",
+        description="The end date of the residence permit validity",
+        examples=[datetime.datetime(2024, 2, 19)],
+    )
+    industries: List[float] = Field(
```
Comment (joakimnordling) [inline src/Permits/WorkPermit_v0.1.py:50]: This was missing the type (just `List[]` in the spreadsheet). The NACE codes should be strings, not floats, so please update to list of strings and remember to update the examples as well.

File: src/Employment/IncomeTax_v0.1.py [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+    )
+    additional_percentage: float = Field(
+        ...,
+        title="Additional Percentage",
+        description="The secondary withholding percentage of the income",
+        examples=[0.45],
+    )
+    validity_period: date = Field(
+        ...,
+        title="Validity Period",
```
Comment (joakimnordling) [inline src/Employment/IncomeTax_v0.1.py:45]: Wrong field name and title.

File: src/Permits/WorkPermit_v0.1.py [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+        title="Employer Name",
+        description="The name of the employer that the person is allowed to work for",
+        examples=["Staffpoint Oy"],
+    )
+
+
+class WorkPermitResponse(CamelCaseModel):
+    permits: List[Permit] = Field(
+        ...,
+        title="Permits",
```
Comment (joakimnordling) [inline src/Permits/WorkPermit_v0.1.py:68]: Missing description. The description was wrong in the spreadsheet, but is now updated there (checked it together with Henna).

File: src/Permits/WorkPermit_v0.1.py [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+        title="Validity End",
+        description="The end date of the residence permit validity",
+        examples=[datetime.datetime(2024, 2, 19)],
+    )
+    industries: List[str] = Field(
+        ...,
+        title="Industries",
+        description="The list of industries that the permit holder is allowed to work based on the Statistical "
+        "classification of economic activities in the European Community, abbreviated as NACE.",
+        examples=["79.1, 79.9"],
```
Comment (joakimnordling) [inline src/Permits/WorkPermit_v0.1.py:55]: I think this is how it should be:
```suggestion
        examples=[["79.1", "79.9"]],
```

**Review by joakimnordling — APPROVED** [review]
